### PR TITLE
add binary memtable

### PR DESCRIPTION
**Pull request by cshuo**
## What is the purpose of the change

Implement a binary mem-table with binary sort

## Brief change log
- Implement a binary mem-table with binary sort

## Verifying this change
- current it case

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/TableStorageOptions.java [hunk 1]
```diff
@@ -100,4 +101,15 @@
...
                     .defaultValue(1)
                     .withDescription("");
+
+    public static final ConfigOption<Boolean> FILE_BINARY_MEM_TABLE_ENABLED =
+            ConfigOptions.key("mem-table.binary-enabled")
+                    .booleanType()
+                    .defaultValue(true)
+                    .withDescription("");
+    public static final ConfigOption<MemorySize> TABLE_EXEC_RESOURCE_EXTERNAL_BUFFER_MEMORY =
+            key("mem-table.memory")
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/TableStorageOptions.java]: file.write-buffer.memory-size

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/TableStorageOptions.java [hunk 2]
```diff
@@ -100,4 +101,15 @@
                     .intType()
                     .defaultValue(1)
                     .withDescription("");
+
+    public static final ConfigOption<Boolean> FILE_BINARY_MEM_TABLE_ENABLED =
+            ConfigOptions.key("mem-table.binary-enabled")
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/TableStorageOptions.java]: file.write-buffer.binary-enabled

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/BinaryMemTable.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                @Override
+                public boolean advanceNext() throws IOException {
+                    switchKeyValue();
+
+                    Tuple2<BinaryRowData, BinaryRowData> kv =
+                            kvIter.next(Tuple2.of(reuseKey, reuseValue));
+                    if (kv == null) {
+                        return false;
+                    }
+                    BinaryRowData userKey = projection.apply(reuseKey).copy();
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/BinaryMemTable.java]: We should not copy key and value.
We Can use something like `OffsetRowData`.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/BinaryMemTable.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                        .map(RowType.RowField::getType)
+                        .collect(Collectors.toList());
+        types.add(new BigIntType());
+        types.add(new TinyIntType());
+        LogicalType[] keyTypes = types.toArray(new LogicalType[0]);
+        LogicalType[] valTypes =
+                valueType.getFields().stream()
+                        .map(RowType.RowField::getType)
+                        .toArray(LogicalType[]::new);
+        this.table = new BytesHashMap(null, null, maxMemSize, keyTypes, valTypes);
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/BinaryMemTable.java]: Intead of using `BytesHashMap`, maybe we can concat them as a Row just like `SstFileWriter`.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/FileStoreImpl.java [hunk 5]
```diff
@@ -86,7 +89,11 @@ public FileStoreImpl(
         this.keyComparator = keyComparator;
         this.writerFactory = writerFactory;
         this.readerFactory = readerFactory;
-        this.memTable = new HeapMemTable(keyComparator);
+        if (options.binaryEnabled) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/FileStoreImpl.java]: use `? : `

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/BinaryMemTable.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        reuseMetaRow.setField(1, valueType.toByteValue());
+        reuseStoreRow.replace(key, reuseMetaRow);
+
+        BytesMap.LookupInfo<BinaryRowData, BinaryRowData> lookupInfo =
+                table.lookup(storeKeySerializer.toBinaryRow(reuseStoreRow));
+        assert !lookupInfo.isFound();
+        try {
+            table.append(lookupInfo, valueSerializer.toBinaryRow(value));
+        } catch (IOException e) {
+            throw new StoreException("MemTable is full, cannot add any more records.");
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/BinaryMemTable.java]: I prefer put must throws `EOFException`.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/FileStoreImpl.java [hunk 7]
```diff
@@ -100,23 +107,29 @@ public FileStoreImpl(
...
 
     @Override
     public void delete(RowData key, RowData value) throws StoreException {
-        memTable.put(levels.newSequenceNumber(), ValueKind.DELETE, key, value);
-        checkFlush();
+        try {
+            memTable.put(levels.newSequenceNumber(), ValueKind.DELETE, key, value);
+        } catch (StoreException e) {
+            doFlush();
+            delete(key, value);
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/FileStoreImpl.java]: `memTable.put(levels.newSequenceNumber(), ValueKind.DELETE, key, value);` and just throws exception ("the write buffer is too small") if fail.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/StoreOptions.java [hunk 8]
```diff
@@ -29,22 +29,28 @@
...
     public final int maxSizeAmplificationPercent;
     public final int sizeRatio;
+    public final boolean binaryEnabled;
 
-    // TODO replace maxMemRecords with maxMemBytes
+    // default maxMemRecords
     public final int maxMemRecords = 10_000;
 
+    // default mem size 32M
+    public final long maxMemBytes = 32_000_000L;
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/StoreOptions.java]: init it by constructor?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/planner/FileStoreFactory.java [hunk 9]
```diff
@@ -71,6 +78,8 @@ public FileStore create(String partition, int bucket, List<SstFileMeta> sstFiles
                 new Path(tablePath, FileFactory.BUCKET_DIR_PREFIX + bucket),
                 keySerializer.getArity(),
                 valueSerializer.getArity(),
+                keyType,
+                valueType,
                 keySerializer,
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/planner/FileStoreFactory.java]: remove key ser and value ser.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/BinaryMemTable.java (hunk 4 above)
Comment (cshuo) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/BinaryMemTable.java]: Yes...all we need here is an sequentially append-only bytes buffer.

**Pull request closed without merging**